### ISSUE TITLE
Error in `torch.export` / `torch.compile` while tracing due to data-dependent control flow in `get_rope_index` (Qwen3-VL, Alpamayo)

### ISSUE THREAD

**Issue by Mani-offl** [pytorch/pytorch#173915]
### Summary
The current implementation of `get_rope_index` contains data-dependent Python control flow (`.item()` and `if` statements on tensor-derived values).  
This causes `torch.export.export` to fail with a data dependent error during tracing.

---

### Affected Function

```python
def get_rope_index(
        self,
        input_ids: torch.LongTensor,
        image_grid_thw: torch.LongTensor | None = None,
        video_grid_thw: torch.LongTensor | None = None,
        attention_mask: torch.Tensor | None = None,
    ):
        device = input_ids.device
        B, S = input_ids.shape

        vision_start_token_id = self.config.vision_start_token_id
        image_token_id = self.config.image_token_id
        video_token_id = self.config.video_token_id
        spatial_merge_size = self.config.vision_config.spatial_merge_size

        if attention_mask is None:
            attention_mask = torch.ones_like(input_ids, dtype=torch.long)

        if image_grid_thw is None and video_grid_thw is None:
            pos = attention_mask.cumsum(-1) - 1
            pos.masked_fill_(attention_mask == 0, 0)
            position_ids = pos.unsqueeze(0).expand(3, -1, -1).contiguous()
            rope_deltas = torch.zeros((B, 1), device=device, dtype=torch.long)
            return position_ids, rope_deltas

        if video_grid_thw is not None and video_grid_thw.numel() > 0:
            video_grid_thw = torch.repeat_interleave(
                video_grid_thw, video_grid_thw[:, 0], dim=0
            )
            video_grid_thw[:, 0] = 1

        position_ids = torch.zeros((3, B, S), device=device, dtype=torch.long)
        rope_deltas = torch.zeros((B, 1), device=device, dtype=torch.long)

        zero_scalar = torch.zeros((), device=device, dtype=torch.long)
        zero_thw = torch.zeros(3, device=device, dtype=torch.long)

        for b in range(B):
            mask = attention_mask[b].bool()
            tokens = input_ids[b]

            masked_indices = torch.nonzero(mask, as_tuple=False).squeeze(1)

            comp_pos = attention_mask[b].to(torch.long).cumsum(0) - 1
            comp_pos = torch.clamp(comp_pos, min=0)

            llm_pos_chunks = []
            st = torch.zeros((), device=device, dtype=torch.long)

            img_ptr = 0
            vid_ptr = 0

            def next_base(chunks):
                if not chunks or chunks[-1].numel() == 0:
                    return torch.zeros((), device=device, dtype=torch.long)
                return chunks[-1].max() + 1

            masked_indices_safe = torch.cat(
                [masked_indices, masked_indices.new_zeros((1,))], dim=0
            )
            num_masked = masked_indices.numel()
            num_masked_t = torch.tensor(num_masked, device=device, dtype=torch.long)
            max_masked_idx = torch.clamp(num_masked_t - 1, min=0)

            for pos in range(S):
                pos_t = comp_pos[pos]
                is_vs = mask[pos] & (tokens[pos] == vision_start_token_id)

                text_len = torch.where(is_vs, pos_t - st, zero_scalar)

                base = next_base(llm_pos_chunks)

                if text_len.item() > 0:
                    llm_pos_chunks.append(
                        torch.arange(text_len, device=device, dtype=torch.long)
                        .view(1, -1)
                        .expand(3, -1)
                        + base
                    )

                base = base + text_len

                next_pos_idx = torch.minimum(pos_t + 1, max_masked_idx)
                next_pos_idx_i = next_pos_idx.to(torch.long).unsqueeze(0)

                next_abs_pos_tensor = torch.gather(
                    masked_indices_safe, 0, next_pos_idx_i
                )
                next_abs_pos = next_abs_pos_tensor[0]
                next_tok = tokens[next_abs_pos]

                is_image = is_vs & (next_tok == image_token_id)
                is_video = is_vs & (next_tok == video_token_id)

                img_thw = (
                    image_grid_thw[img_ptr]
                    if image_grid_thw is not None and img_ptr < image_grid_thw.shape[0]
                    else zero_thw
                )
                vid_thw = (
                    video_grid_thw[vid_ptr]
                    if video_grid_thw is not None and vid_ptr < video_grid_thw.shape[0]
                    else zero_thw
                )

                thw = torch.where(is_image, img_thw, vid_thw)
                t, h, w = thw

                gh = h // spatial_merge_size
                gw = w // spatial_merge_size
                vision_len = gh * gw

                if vision_len.item() > 0:
                    t_idx = torch.zeros(vision_len, device=device, dtype=torch.long)
                    h_idx = torch.arange(gh, device=device).repeat_interleave(gw)
                    w_idx = torch.arange(gw, device=device).repeat(gh)

                    llm_pos_chunks.append(
                        torch.stack([t_idx, h_idx, w_idx]) + base
                    )

                img_ptr += int(is_image.item())
                vid_ptr += int(is_video.item())

                st = torch.where(is_vs, pos_t + vision_len, st)

            base = next_base(llm_pos_chunks)
            tail_len = num_masked - int(st)

            if tail_len > 0:
                llm_pos_chunks.append(
                    torch.arange(tail_len, device=device, dtype=torch.long)
                    .view(1, -1)
                    .expand(3, -1)
                    + base
                )

            if not llm_pos_chunks:
                llm_positions = torch.zeros((3, 0), device=device, dtype=torch.long)
            else:
                llm_positions = torch.cat(llm_pos_chunks, dim=1)

            position_ids[:, b, masked_indices] = llm_positions

            if llm_positions.numel() == 0:
                rope_deltas[b, 0] = 0
            else:
                rope_deltas[b, 0] = llm_positions.max() + 1 - num_masked

        return position_ids, rope_deltas
```

### Problematic code
```
if text_len.item() > 0:
                    llm_pos_chunks.append(
                        torch.arange(text_len, device=device, dtype=torch.long)
                        .view(1, -1)
                        .expand(3, -1)
                        + base
                    )
```
---

### Error Observed

When calling `torch.export.export`:
```
torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode:
Could not extract specialized integer from data-dependent expression u163
text_len.item() > 0
```
The same error is observed when running the model with `torch.compile`, even if we set the flag - `torch._dynamo.config.capture_scalar_outputs = True` as mentioned in the issue #114483
```
Graph Break Reason: Could not extract specialized integer from data-dependent expression
```
---

**Has anyone encountered this issue before?**
**Are there recommended ways to refactor this pattern to make it compatible with `torch.export` / `torch.compile`, or any known workarounds?**

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo

**Comment by Mani-offl** [comment]
Thanks for the reference. I went through the programming model section.

In my case, the condition `(text_len > 0)` is computed **inside an intermediate layer**, not at the model boundary. Because of that, moving the condition *outside tracing* isn’t feasible without introducing a graph break or restructuring the model.

```
# --------------ORIGINAL CODE---------------------------
if text_len.item() > 0:
    llm_pos_chunks.append(
        torch.arange(text_len, device=device, dtype=torch.long)
        .view(1, -1)
        .expand(3, -1)
        
    )

# --------------MODIFIED CODE AS PER THE DOC---------------------------
cond = (text_len > 0).item()

@torch.compile
if cond:
    llm_pos_chunks.append(
        torch.arange(text_len, device=device, dtype=torch.long)
        .view(1, -1)
        .expand(3, -1)
        
    )
```

I also tried torch.cond(), but here the condition affects padding and the overall data flow, which ends up changing the model outputs. Since this same .item() pattern is used in many places, fixing it through padding or branch alignment would be very time-consuming.

Is there any other workaround you’d recommend for handling this kind of data-dependent logic?

**Comment by Mani-offl** [comment]
Thanks for the breakdown.

The main issue was the `.item()` dependency, which I was able to resolve by replacing the problematic code with:

Issue:

```python
 if text_len.item() > 0:
                    llm_pos_chunks.append(
                        torch.arange(text_len, device=device, dtype=torch.long)
                        .view(1, -1)
                        .expand(3, -1)
                        + base
                    )
```

Workaround:
```python
from torch.fx.experimental.symbolic_shapes import guard_or_false
chunk = torch.arange(text_len, device=device, dtype=torch.long).view(1, -1).expand(3, -1) + base
 if guard_or_false(chunk.numel() > 0):
          ...

```

And as you mentioned I was able to rewrite the code for the issues caused by the python list dependency and dynamic pointers - imp_ptr, vid_ptr and also few others.

Regarding the `for pos in range(S):` loop — since `S` is derived from `input_ids.shape`, it is treated as a static dimension or a symbolic shape rather than a value calculated dynamically from tensor data. As it is not computed at run time, it remains trace-friendly and didn't pose an issue for the export process.

After the changes i was able to export and compile the mode without any graph breaks.
Appreciate the help and insights. I’m closing this issue for now — please feel free to reopen if further discussion is needed.

Thanks again!
